# 💻 Question No: 05

## ⚙️ Setup

- Ensure the Python kernel has the necessary libraries: `pandas`, `matplotlib` and `lets-plot`,`os`,`numpy`,`statsmodels`,`seaborn`
- Ensure the `bank-full.csv` file is in the `data` folder.

In [1]:
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.metrics.pairwise import cosine_similarity
import seaborn as sns
import os
from sklearn import tree
os.getcwd()
import numpy as np
import statsmodels.api as sm

from lets_plot import * # This imports all of ggplot2's functions
LetsPlot.setup_html()

In [2]:
df = pd.read_excel('D:/Data Science for Marketing-I/data/Online Retail.xlsx')
df 

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[541909 rows x 8 columns]

### Exclude entries where "Quantity" or "UnitPrice" have negative or zero values, and remove observations with missing CustomerID.



In [3]:

# Filter data
df = df[(df['Quantity'] > 0) & (df['UnitPrice'] > 0)]
df = df.dropna(subset=['CustomerID'])

💡df will contain only rows where:
Quantity and UnitPrice are both positive.
CustomerID is not missing

In [4]:
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[397884 rows x 8 columns]

### Create a Customer-Item Matrix using the pivot table function, replacing NaN values with 0 and non-NaN values with 1.

In [5]:
# Create Customer-Item Matrix
customer_item_matrix = df.pivot_table(index='CustomerID', columns='StockCode', values='Quantity', aggfunc='sum')
customer_item_matrix = (customer_item_matrix > 0).astype(int)
customer_item_matrix.head()



StockCode   10002  10080  10120  10125  10133  10135  11001  15030  15034  \
CustomerID                                                                  
12346.0         0      0      0      0      0      0      0      0      0   
12347.0         0      0      0      0      0      0      0      0      0   
12348.0         0      0      0      0      0      0      0      0      0   
12349.0         0      0      0      0      0      0      0      0      0   
12350.0         0      0      0      0      0      0      0      0      0   

StockCode   15036  ...  90214V  90214W  90214Y  90214Z  BANK CHARGES  C2  DOT  \
CustomerID         ...                                                          
12346.0         0  ...       0       0       0       0             0   0    0   
12347.0         0  ...       0       0       0       0             0   0    0   
12348.0         0  ...       0       0       0       0             0   0    0   
12349.0         0  ...       0       0       0       0             0   0    0   
12350.0         0  ...       0       0       0       0             0   0    0   

StockCode   M  PADS  POST  
CustomerID                 
12346.0     0     0     0  
12347.0     0     0     0  
12348.0     0     0     1  
12349.0     0     0     1  
12350.0     0     0     1  

[5 rows x 3665 columns]

💡Rows (Index): Represent unique CustomerIDs (e.g., 12346.0, 12347.0, etc.).
Columns: Represent unique StockCodes (e.g., 10002, 10080, etc.), which are product codes.
Values: Likely represent the count or quantity of each product (StockCode) purchased by each customer (CustomerID).

In [6]:
# Compute User-to-User Similarity Matrix
similarity_matrix = pd.DataFrame(cosine_similarity(customer_item_matrix), 
                                 index=customer_item_matrix.index, 
                                 columns=customer_item_matrix.index)
similarity_matrix.head()

CustomerID  12346.0   12347.0   12348.0   12349.0   12350.0   12352.0  \
CustomerID                                                              
12346.0         1.0  0.000000  0.000000  0.000000  0.000000  0.000000   
12347.0         0.0  1.000000  0.063022  0.046130  0.047795  0.038484   
12348.0         0.0  0.063022  1.000000  0.024953  0.051709  0.027756   
12349.0         0.0  0.046130  0.024953  1.000000  0.056773  0.137137   
12350.0         0.0  0.047795  0.051709  0.056773  1.000000  0.031575   

CustomerID  12353.0   12354.0   12355.0   12356.0  ...  18273.0   18274.0  \
CustomerID                                         ...                      
12346.0         0.0  0.000000  0.000000  0.000000  ...      0.0  0.000000   
12347.0         0.0  0.025876  0.136641  0.094742  ...      0.0  0.029709   
12348.0         0.0  0.027995  0.118262  0.146427  ...      0.0  0.064282   
12349.0         0.0  0.030737  0.032461  0.144692  ...      0.0  0.105868   
12350.0         0.0  0.000000  0.000000  0.033315  ...      0.0  0.000000   

CustomerID   18276.0  18277.0   18278.0   18280.0  18281.0   18282.0  \
CustomerID                                                             
12346.0     0.000000      0.0  0.000000  0.000000      0.0  0.000000   
12347.0     0.052668      0.0  0.032844  0.062318      0.0  0.113776   
12348.0     0.113961      0.0  0.000000  0.000000      0.0  0.000000   
12349.0     0.000000      0.0  0.039014  0.000000      0.0  0.067574   
12350.0     0.000000      0.0  0.000000  0.000000      0.0  0.000000   

CustomerID   18283.0   18287.0  
CustomerID                      
12346.0     0.000000  0.000000  
12347.0     0.109364  0.012828  
12348.0     0.170905  0.083269  
12349.0     0.137124  0.030475  
12350.0     0.044866  0.000000  

[5 rows x 4338 columns]

💡CustomerID 12346.0 has no similarity with any other customer (all values are 0.0 except for itself).

CustomerID 12347.0 has some similarity with other customers, such as 12348.0 (0.063022) and 12349.0 (0.046130).

CustomerID 12348.0 has a higher similarity with 18283.0 (0.170905), indicating they have more similar purchasing behavior.

### Compute the User-to-User Similarity Matrix.
Recommend products to the user who has the highest similarity to customer 17173.


In [7]:
most_similar_user = similarity_matrix.loc[17173].sort_values(ascending=False).index[1]

# Recommend products
customer_17173_items = set(customer_item_matrix.loc[17173][customer_item_matrix.loc[17173] > 0].index)
most_similar_user_items = set(customer_item_matrix.loc[most_similar_user][customer_item_matrix.loc[most_similar_user] > 0].index)
recommended_items = most_similar_user_items - customer_17173_items

print(recommended_items)


{22568, 23128}


💡 This indicates products with StockCodes 85099B and 84406B are recommended to customer 17173.

### Additionally, apply item-based collaborative filtering to identify products similar to the item with stock code 90103

In [8]:
# Item-Based Collaborative Filtering
item_similarity_matrix = pd.DataFrame(cosine_similarity(customer_item_matrix.T), 
                                      index=customer_item_matrix.columns, 
                                      columns=customer_item_matrix.columns)

similar_items_to_90103 = item_similarity_matrix[90103].sort_values(ascending=False).head(6).index[1:]
print(similar_items_to_90103)


Index(['90059B', '90059E', '90059F', 90101, '90059C'], dtype='object', name='StockCode')


💡These are the product codes (StockCode) that the most similar customer has purchased but customer 17173 has not.

The recommendations include:
'90059B'
'90059E'
'90059F'
90101
'90059C'

In [9]:
df.loc[df['StockCode'].isin(similar_items_to_90103),['StockCode','Description']
       ].drop_duplicates().set_index('StockCode')

Description
StockCode                                     
90059B     DIAMANTE HAIR GRIP PACK/2 BLACK DIA
90059E          DIAMANTE HAIR GRIP PACK/2 RUBY
90059C       DIAMANTE HAIR GRIP PACK/2 MONTANA
90059F       DIAMANTE HAIR GRIP PACK/2 LT ROSE
90101                WHITE FRANGIPANI NECKLACE

💡 

Finds users with similar purchase behavior to recommend relevant products.
Insights:
The first four products are hair accessories with slight variations in color/design.
The fifth product is a necklace, indicating potential interest in fashion-related items.